File: haskell/p002.hs
fibs :: [Int]
fibs = 1 : 2 : [a + b | (a, b) <- zip fibs (tail fibs)]

isEven :: Int -> Bool
isEven n = (mod n 2) == 0

main = print (sum (takeWhile (< 4000000) (filter isEven fibs)))

File: scala/src/P002.scala
object P002 {
  def main(args: Array[String]) {
    lazy val fibs: Stream[BigInt] = BigInt(1) #:: BigInt(2) #:: fibs.zip(fibs.tail).map { n => n._1 + n._2 }
    println(fibs.takeWhile(_ < 4000001).filter { _ % 2 == 0 }.sum)
  }
}
